### PR TITLE
rfc: introduce logsigmoid support

### PR DESCRIPTION
**Pull request by piotrchmiel**
[Rendered document](https://github.com/piotrchmiel/oneDNN/tree/logsigmoid/rfcs/20200609-logsigmoid)

RFC concerns introduction of logsigmoid op support in oneDNN. Full motivation is described inside document.

### CODE REVIEW

File: rfcs/20200609-logsigmoid/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Introducing logsigmoid operation
+
+## Introduction
+
+Request comes from DGP graph lib community https://www.dgl.ai. Logsigmoid is
+used in AWS embedding models https://bit.ly/30kWI1f. Currently logsigmoid can be
+implemented with oneDNN by using two eltwise
+(https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html) operations
+logistic and log. Unfortunately current eltwise ops are not fusable.
```
Comment (rsdubtso) [inline rfcs/20200609-logsigmoid/README.md:9]: Can you please provide formulas for forward and backward (including backward computed via dst only)?

This may help decide whether we need a separate algorithm or not.

File: rfcs/20200609-logsigmoid/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+Pros:
+- reuse existing implementation internally;
+- flexibility - can be used each time when there will be need for fuse existing
+ops;
+
+Cons:
+- more complicated in implementation than option 1;
+
+I recommend implementing option 2, because of better flexibility.
```
Comment (dzarukin) [inline rfcs/20200609-logsigmoid/README.md:58]: Before go with option 2, the question of general backward applicability should be explored: is it possible to apply backward computations in reversed order without spoiling the answer? If yes, how do we expect user to pass us attributes for backward: from hint or pass them explicitly? How does it work with `use_dst_for_bwd` algorithms? How does it work in case two kinds are mixed: one which supports dst on bwd, and the other one does not?

Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:58]: Let me think: 
In single eltwise case for op named a, forward is defined:
d= f(s), 
so backward is 
ds = f'(s) * dd, where f'(s) is derivative, and dd input gradient computed on the preceding layer. 
If we fuse two eltwise ops named a and b (b(a)) forward is defined as
d = g(f(s)), 
so backward is 
ds = g(f(s))' * dd
General rule is:  g(f(s))' = g'(x) * f'(g(x)),
so ds =  g'(x) * f'(g(x)) * dd.
So, if your understanding of reversed order calculations in backward is that we take eltwise ops a, b, take calculated derivatives for single op a_der = f'(x), b_der =g'(x), and than apply in backward is ds = b_der(a_der(x)) *dd,  it seems that you're right and general backward applicability is not that easy. Could you please check if I didn't made any logical mistake ?

File: rfcs/20200609-logsigmoid/README.md (hunk 1 above)
Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:9]: I updated RFC with the formulas

File: rfcs/20200609-logsigmoid/README.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+loads logistic output from memory and stores data (2x loads and 2x stores). While
+it could be possibly 1x load and 1x store if these operations were combined.
+Eltwise is memory bound primitive, so additional loads and store has performance
+impact.
+
+Naming convention taken from: https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html
+FWD derivative calculated using https://www.wolframalpha.com/input/?i=d%2Fdx%28log%281%2F%281%2Bexp%28-x%29%29%29%29
+
+Formulas:
+- forward_logsigmoid - d = log(1/(1+exp(-x)));
```
Comment (emfomenk) [inline rfcs/20200609-logsigmoid/README.md:20]: I wonder if there are any expectations on the accuracy of the logsigmoid. We might need to check that with framework guys. I doubt, but it might happen that they consider this complex operation as a whole, that (in theory) could give better accuracy, than chaining it. For instance:
```
log(1/1+e^-x)) = -log(1 + e^-x)
```

If it turns out they use something like this, I doubt we have an option to use just post-ops.

File: rfcs/20200609-logsigmoid/README.md (hunk 2 above)
Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:66]: Just a note: if the focus is on inference now, that's probably not that big of deal to skip backward (meaning that for training the integration implementation will have to split the operation in two).

On the other hand, if we want training and, say bf16, we might really want to have a alg_kind, as this will allow use to have f32 data type through out the who backward computations, while separating the operation in two will most likely lead to bf16 intermediate result. Probably worth mentioning in pros and cons of the options.

Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:66]: so far I incline towards option 1 as the one that (potentially) addresses most of the issues

File: rfcs/20200609-logsigmoid/README.md (hunk 3 above)
Comment (rsdubtso) [reply rfcs/20200609-logsigmoid/README.md:20]: Nit:

>```
> log(1/1+e^-x)) = -log(1 + e^-x)
> ```

I think you meant
```
log(1/(1+e^(-x))) = -log(1 + e^(-x))
```

File: rfcs/20200609-logsigmoid/README.md (hunk 2 above)
Comment (rsdubtso) [reply rfcs/20200609-logsigmoid/README.md:66]: I second option 1.

But I have some additional comments...

> say bf16, we might really want to have a alg_kind, as this will allow use to have f32 data type through out the whole backward computations

I'm not sure understand how fusion prevents you from having the whole backward computed in `f32`. It is would probably be challenging to do this in the current API, but eventually I think our eltwise should provide just very basic functions which could be used together to compute more complex expressions. This will also make it possible to forego all the backward and forward and just allow users to spell out the required expressions.

File: rfcs/20200609-logsigmoid/README.md (hunk 1 above)
Comment (dsciebu) [reply rfcs/20200609-logsigmoid/README.md:9]: In case you are describing, this would demand combining eltwise linear with eltwise soft_relu (or opposite?) for backward, correct? How to combine those ops then? It doesn't seem to give the same output formula as combining logistic followed by log as @piotrchmiel proposed.

File: rfcs/20200609-logsigmoid/README.md (hunk 2 above)
Comment (dsciebu) [reply rfcs/20200609-logsigmoid/README.md:66]: @piotrchmiel I think @dzarukin touched a very important topic - depending on fuses combination, it might be very important (from perf view) to reuse intermediate computation results (done in fwd pass) to calculate bwd. In such case, how should user set up those arguments? If we assume that the user can fuse any number of eltwise ops in any order how does the mechanism of passing args should look like? Or maybe, idea is there is no intermediates reuse and some computations need to be repeated if their results are needed? This in turn may be very complicated.

Comment (rsdubtso) [reply rfcs/20200609-logsigmoid/README.md:66]: My idea is to *not* have any forward or backward operations, just `+`, `-`, `*`, `exp(x)`, ... etc. This is not proposed to be done *now*, but this is where I think it makes sense to go. This would require major overhaul of our fusion capabilities too. I'm resolving this discussion I seem to be causing more confusion than helping.

File: rfcs/20200609-logsigmoid/README.md (hunk 3 above)
Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:20]: fixed, thx!

Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:20]: Mentioned formula is used in tensorflow https://www.tensorflow.org/api_docs/python/tf/math/log_sigmoid (For numerical stability, we use y = -tf.nn.softplus(-x)). In other framworks seems that original formula of log_sigmoid is used. Anyway I will ask people who requested feature.

Comment (rongzha1) [reply rfcs/20200609-logsigmoid/README.md:20]: Hi ,post-ops may result in numerical unstability, in pytorch also use following trick
```c
    scalar_t max_elem = fmax(0, -*input_data);
    scalar_t z = exp(-max_elem) + exp(-*input_data - max_elem);
    *buffer_data = z;
    *output_data = -(max_elem + log(z));
```
https://github.com/pytorch/pytorch/blob/72e17715dd7e97af518b3e557ded646c23b62e90/aten/src/THNN/generic/LogSigmoid.c#L13

Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:20]: So, the conclusions are as follows:
- we are more inclined towards extending list of operations instead of using post-ops;
- w don't want to use `log(1/(1+e^-x))`, because of it may result in numerical unstability;
Open question:
- Which formula we should use for numerical stability tf or pytorch ?

Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:20]: That's a good question.
The formula shown by @rongzha1 is protected against overflows, and we do similar trick in (log)softmax. The problem, however, that this approach is no longer an element-wise in some sense -- you need to gather the statistic over the whole buffer. That would prevent the possibility of fusion and will cost us performance.
So, IMO if we implement `logsigmoid`, we implement it straightforward (i.e. TF version)...

@rongzha1, do you think that would be a no-go for PyTorch integration?
In theory, we could have the following work arounds:
1. Use `alpha` parameter for `max_elem`. It won't be computed at run time, but if really required, users will have a know to improve the stability (manually). This is ugly and I really don't think that path forward. But still.
2. Specifically for PyTorch implement `logsigmoid_accurate`, but state in the documentation that this alg_kind cannot be used for fused computations.

Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:20]: @piotrchmiel, please add this aspect to the RFC as well.
Also, could you please check what does MxNet use?

Comment (rongzha1) [reply rfcs/20200609-logsigmoid/README.md:20]: OK. both are LGTM.  In fact, I'm not working on framework integration.

Comment (TaoLv) [reply rfcs/20200609-logsigmoid/README.md:20]: @emfomenk AFAIK, MXNet doesn't have a standalone logsigmoid operator. That means usually it's implemented via `log(sigmoide(x))` in model. Once this primitive is ready in oneDNN library, we can add the corresponding operator to  framework.

Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:20]: Thank you, @TaoLv and @romgzha1!

Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:20]: @emfomenk Why do you think that we need to gather statistics over the whole buffer when using pytorch approach ? Please correct my if I'm wrong, but based on proposed workaround 1, where you suggested to pass max_elem, as alpha, it suggest that we have one max_elem per entire tensor. But if I didn't missed anything in pytorch code https://github.com/pytorch/pytorch/blob/72e17715dd7e97af518b3e557ded646c23b62e90/aten/src/THNN/generic/LogSigmoid.c#L13 - logsigmoid code is wrapped by macro TH_TENSOR_APPLY3, which is for loop over all elements. So computation `scalar_t max_elem = fmax(0, -*input_data);` is per element not per entire tensor. As it is per element, I don't understand why it is not elementwise in some sense ? It is still single pass over the data.

Comment (emfomenk) [reply rfcs/20200609-logsigmoid/README.md:20]: Hi @piotrchmiel,

I don't know which part of the body I used to think when wrote that (but that definitely wasn't brain).
Please disregard my message. You are right, this is not per-tensor but per-element maximum.

I still think we should start with the straightforward formula though :)

Resolving the thread. Approving the RFC.
Thank you and sorry for the noise.

File: rfcs/20200609-logsigmoid/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Cons:
+- flexibility: extending op list each time when exisiting can be fused seems to
+be poor general solution;
+
+**Option 2 Add support for eltwise post-ops in eltwise primitive**
+
+``` cpp
+    auto eltwise_d = eltwise_forward::desc(prop_kind::forward_training,
+            algorithm::dnnl_eltwise_logistic, src_md, 0.f, 0.f);
```
Comment (igorsafo) [inline rfcs/20200609-logsigmoid/README.md:53]: I
```suggestion
            algorithm::eltwise_log, src_md, 0.f, 0.f);
```

Comment (piotrchmiel) [reply rfcs/20200609-logsigmoid/README.md:53]: Why logistic should be change to log ?

**Comment by TaoLv** [comment]
Do we have any planed release to have this feature? @piotrchmiel @vpirogov

File: rfcs/20200609-logsigmoid/README.md (hunk 4 above)
Comment (igorsafo) [reply rfcs/20200609-logsigmoid/README.md:53]: @piotrchmiel fixed, i meant that data type from C++ API should be used here
